### PR TITLE
[markdown] Harder, better, faster, stronger `vite-plugin-markdown`

### PR DESCRIPTION
**Pull request by bholmesdev**
## Changes

- Rework `vite-plugin-markdown` to 1 compilation step
  - No more Astro + esbuild compilation, no more 2 step imports, no more mysterious `?content` and `?mdImport` flags 🥳 
  - Pass through `renderMarkdown` in all cases, including globs - unblocks remark / rehype plugins appending to frontmatter
  - Use Astro's JSX runtime to render markdown layouts
  - Avoid switch logic on `legacy.astroFlavoredMarkdown`
- Fall back to old `vite-plugin-markdown` (now renamed `vite-plugin-markdown-legacy`) based on legacy flag
- Enable JSX runtime in all cases, not just MDX!
  - Tweak JSX renderer logic to better handle Astro renderer (based on failing unit tests)
- Add `headings` and `frontmatter` to layout props, for consistency with MDX (#4134)

## Testing

- move `astro-markdown` suite to `legacy-astro-flavored-markdown` for clarity
- port legacy tests to `astro-markdown` suite (formerly `astro-markdown-md-mode`): layouts, headings, `import.meta.env`, raw and compiled content

## Docs

N/A

### CODE REVIEW

File: packages/astro/src/vite-plugin-markdown-legacy/index.ts
Comment (FredKSchott) [inline packages/astro/src/vite-plugin-markdown-legacy/index.ts:145]: no longer needed! It would be great to remove these branching logic flows from the plugin code now that it's flipped on/off outside of this plugin

Comment (bholmesdev) [reply packages/astro/src/vite-plugin-markdown-legacy/index.ts:145]: Fair enough! Just left it untouched because we plan to delete this module anyways (with no future changes). I can refactor though.

Comment (bholmesdev) [reply packages/astro/src/vite-plugin-markdown-legacy/index.ts]: Done 👍

File: examples/docs/src/components/PageContent/PageContent.astro
Comment (bholmesdev) [inline examples/docs/src/components/PageContent/PageContent.astro:7]: `content.astro.headings` now throws! You'll get a nice error recommending `Astro.props.headings` instead 👍

File: packages/astro/src/core/build/graph.ts
Comment (bholmesdev) [inline packages/astro/src/core/build/graph.ts:32]: This assumed a `pages` entry would never be imported, which breaks `glob` imports for MD files in `/pages`. The `?mdImport` flag hid this bug before!

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (FredKSchott) [inline packages/astro/src/vite-plugin-markdown/index.ts:33]: why do we want to escape `import.meta.env` ourselves instead of letting Vite do it?

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (FredKSchott) [inline packages/astro/src/vite-plugin-markdown/index.ts:51]: todo for now or later? just checking that this wasn't a comment to yourself :)

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (bholmesdev) [reply packages/astro/src/vite-plugin-markdown/index.ts:33]: Because Vite does not escape them! It'll replace all `import.meta.env` usages with their values, since Vite assumes all modules are JS modules. A manual file read is the only way around it

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (bholmesdev) [reply packages/astro/src/vite-plugin-markdown/index.ts:51]: AH good catch! Forgot to remove when I addressed the TODO. Removing 👍

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (FredKSchott) [inline packages/astro/src/vite-plugin-markdown/index.ts:92]: I just realized that you're re-calling `JSON.stringify` on the same html/raw values multiple times per file which probably does have a perf impact! 

You could either move these to variables in the plugin context, so that you're running `JSON.stringify` once but inlining the string multiple times in the file itself. OR, you could move these into top-level variables in the file itself, so that `JSON.stringify` is only called once AND the string only exists once in the file.

I'd love for you to explore the later one. I bet my pride that this has a noticeable impact on the DDT benchmark!

**Review by FredKSchott — COMMENTED** [review]
LGTM overall, but left one comment on how to improve the output to reduce potentially-expensive `JSON.stringify` calls.

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (FredKSchott) [reply packages/astro/src/vite-plugin-markdown/index.ts:33]: Ohhhh right, we want them escaped. Makes sense! This is probably more accurate from a Rollup/Vite perspective as well, since I think `load()` was designed to load the file from disk and turn it into JS, while `transform()` is just for transforming the file without changing its file type.

File: packages/astro/src/core/build/graph.ts
Comment (matthewp) [reply packages/astro/src/core/build/graph.ts:32]: I fixed this a different way in https://github.com/withastro/astro/pull/4156

File: packages/astro/src/core/config.ts
Comment (matthewp) [inline packages/astro/src/core/config.ts:347]: why?

File: packages/astro/src/vite-plugin-jsx/index.ts
Comment (matthewp) [inline packages/astro/src/vite-plugin-jsx/index.ts:143]: Would leave this to @natemoo-re but I'm wondering if the astro renderer is intended to be used outside of of just MDX. Otherwise why would it be in astro core and not in the mdx package?

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (matthewp) [inline packages/astro/src/vite-plugin-markdown/index.ts:37]: nit, but adding `, 'utf-8')` to this function call will give you a string.

File: packages/astro/src/vite-plugin-jsx/index.ts
Comment (matthewp) [reply packages/astro/src/vite-plugin-jsx/index.ts:143]: It might be nice to have the jsx renderer API allow you to associate a file extension with a jsx renderer. That way mdx can say that it handles .mdx files while leaving tsx/jsx untouched.

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (natemoo-re) [reply packages/astro/src/vite-plugin-markdown/index.ts:37]: TIL. Nice!

File: packages/astro/src/vite-plugin-jsx/index.ts
Comment (natemoo-re) [reply packages/astro/src/vite-plugin-jsx/index.ts:143]: The Astro JSX renderer certainly _could_ be used for `.[jt]sx` files in the future, but Ben and I discussed this approach as temporary perf shortcut that avoids import scanning for `.mdx` and `.md` files since we already know what runtime they should use.

I like the idea of opening up the JSX renderer API to support explicit file extensions in the future! Sounds like a great v1.1.0 enhancement.

File: packages/astro/src/core/config.ts
Comment (natemoo-re) [reply packages/astro/src/core/config.ts:347]: Previously the Astro JSX renderer was only enabled if `@astrojs/mdx` was included. Now our default `.md` compilation relies on this renderer for Layouts, too.

I don't love that this puts _every_ project in a multiple-JSX renderer world, but I can't think of another option at the moment.

**Review by natemoo-re — APPROVED** [review]
LGTM assuming all other feedback gets addressed! Great work on this, Ben!

File: packages/astro/src/core/build/graph.ts
Comment (bholmesdev) [reply packages/astro/src/core/build/graph.ts:32]: Oh, glad to hear it! Should I revert and let your PR go first @matthewp ?

File: packages/astro/src/vite-plugin-markdown/index.ts
Comment (bholmesdev) [reply packages/astro/src/vite-plugin-markdown/index.ts]: Made a top-level variable! 👍

File: packages/astro/src/core/build/graph.ts
Comment (bholmesdev) [reply packages/astro/src/core/build/graph.ts]: Update: Rebased with your fix 🙏

Comment (matthewp) [reply packages/astro/src/core/build/graph.ts]: Yep, sorry I forgot to reply here.

**Pull request merged**
